contracts/ERC1155NftCustom.sol

```diff
@@ -0,0 +1,188 @@
+            _tokenURIs[_tokenId] = _newUri;
+            emit URI(_newUri, _tokenId);
+        }
+        if (_isFreezeTokenUri) {
+            freezeTokenUris[_tokenId] = true;
+            emit PermanentURI(_tokenURIs[_tokenId], _tokenId);
+        }
+    }
+
+    function burn(
+        uint256 id,
+        uint256 value
+    ) public onlyRole(MINTER_ROLE) {
+        require(tokensBurnable, "NFT: tokens burning is disabled");
+        
+        _burn(_owner, id, value);
+        tokenSupply[id] -= value;
+    }
+
+    function burnBatch(
+        uint256[] memory ids,
+        uint256[] memory values
+    ) public onlyRole(MINTER_ROLE) {
+        require(tokensBurnable, "NFT: tokens burning is disabled");
+        _burnBatch(_owner, ids, values);
+        for (uint256 i = 0; i < ids.length; i++) {
+            tokenSupply[ids[i]] -= values[i];
+        }
+    }
+
+    function transferByOwner(
+        address to,
+        uint256 id,
+        uint256 amount
+    ) public onlyRole(MINTER_ROLE) {
+        require(tokensTransferable, "NFT: Transfers by owner are disabled");
+        _safeTransferFrom(_owner, to, id, amount, "");
+    }
+
+    function transferByOwnerBatch(
+        address[] memory to,
+        uint256[] memory ids,
+        uint256[] memory amounts
+    ) public onlyRole(MINTER_ROLE) {
+        require(tokensTransferable, "NFT: Transfers by owner are disabled");
+        for (uint256 i = 0; i < ids.length; i++) {
+            _safeTransferFrom(_owner, to[i], ids[i], amounts[i], "");
+        }
+    }
+
+    function update(
+        string memory _newBaseURI, 
+        bool _tokensTransferable,
+        bool _freezeUpdates
+    ) public onlyRole(MINTER_ROLE) {
+        require(metadataUpdatable, "NFT: Contract updates are frozen");
+        baseURI = _newBaseURI;
+        if (!_tokensTransferable) {
+            tokensTransferable = false;
+        }
+        if (_freezeUpdates) {
+            metadataUpdatable = false;
+            emit PermanentURIGlobal();
+        }
+    }
+
+    function totalSupply (uint256 _id) public view returns (uint256) {
+        return tokenSupply[_id];
+    }
+
+    function uri(uint256 _id) public override view returns (string memory) {
+        if (bytes(_tokenURIs[_id]).length > 0) {
+            if (bytes(baseURI).length > 0) {
+                return string(abi.encodePacked(baseURI, _tokenURIs[_id]));
+            } else {
+                return _tokenURIs[_id];
+            }
+        } else {
+            return super.uri(_id);
+        }
+    }
+
+    function mintByOwner( address account, uint256 id, uint256 amount, string memory uri) public onlyRole(MINTER_ROLE) {
+        require(!_exists(id), "NFT: token already minted");
+        if (bytes(uri).length > 0) {
+            _tokenURIs[id] = uri;
+            emit URI(uri, id);
+        }
+        _mint(account, id, amount, "");
```

Comment:
Unlikely to be a major issue in this case but external interactions should happen last.